src/Nethermind/Nethermind.Merge.Plugin.Test/InvalidChainTrackerTest.cs

```diff
@@ -0,0 +1,247 @@
+    {
+        Keccak? lastValidHash;
+        List<Keccak> mainChain = MakeChain(20, connectInReverse);
+        List<Keccak> branchAt5 = MakeChain(10, connectInReverse);
+        List<Keccak> branchAt10 = MakeChain(10, connectInReverse);
+        List<Keccak> branchAt15 = MakeChain(10, connectInReverse);
+        List<Keccak> branchAt15_butConnectOnItem5 = MakeChain(10, connectInReverse);
+        List<Keccak> branchAt11_butConnectLater = MakeChain(10, connectInReverse);
+        List<Keccak> branchAt5_butConnectLater = MakeChain(10, connectInReverse);
+        
+        _tracker.SetChildParent(mainChain[1], mainChain[0]);
+        _tracker.SetChildParent(mainChain[1], mainChain[0]);
+        
+        _tracker.SetChildParent(branchAt5[0], mainChain[5]);
+        _tracker.SetChildParent(branchAt10[0], mainChain[10]);
+        _tracker.SetChildParent(branchAt15[0], mainChain[15]);
+        _tracker.SetChildParent(branchAt15_butConnectOnItem5[5], mainChain[15]);
+        
+        _tracker.OnInvalidBlock(mainChain[10], mainChain[9]);
+        
+        _tracker.IsOnKnownInvalidChain(branchAt5[5], out lastValidHash).Should().BeFalse();
+        
+        _tracker.IsOnKnownInvalidChain(branchAt10[0], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        _tracker.IsOnKnownInvalidChain(branchAt10[5], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        
+        _tracker.IsOnKnownInvalidChain(branchAt15[5], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        _tracker.IsOnKnownInvalidChain(branchAt15[9], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        
+        _tracker.IsOnKnownInvalidChain(branchAt15_butConnectOnItem5[9], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        _tracker.IsOnKnownInvalidChain(branchAt15_butConnectOnItem5[5], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        
+        _tracker.IsOnKnownInvalidChain(branchAt15_butConnectOnItem5[4], out lastValidHash).Should().BeFalse();
+        
+        _tracker.SetChildParent(branchAt11_butConnectLater[0], mainChain[11]);
+        _tracker.IsOnKnownInvalidChain(branchAt11_butConnectLater[0], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        _tracker.IsOnKnownInvalidChain(branchAt11_butConnectLater[9], out lastValidHash).Should().BeTrue();
+        lastValidHash.Should().BeEquivalentTo(mainChain[9]);
+        
+        _tracker.SetChildParent(branchAt5_butConnectLater[0], mainChain[5]);
+        _tracker.IsOnKnownInvalidChain(branchAt5_butConnectLater[9], out lastValidHash).Should().BeFalse();
+    }
```

Comment:
Maybe make it  TestCaseSource and split it into multiple Tests that are dependent on previous tests for setting up state? Might be more readable.